### DISCUSSION TITLE
Iterate Query using Iterations

### DISCUSSION THREAD

**Discussion by ccthis** [microsoft/MIMWAL#116] [Ideas]
How can I iterate a query using iteration? I want to close all open/active pam requests (1+) related to the role (1) for each user (1+) that was removed as a candidate from the role. In other words, when one or many users are removed as candidates from a role, the child active requests are forcibly closed. Iteration does not appear to "traverse" back into the query; I get a wal error stating the query is not good showing a Null or blank result when trying to use [//Value] in the query.

This would be a fantastic feature and would really help secure this all 💯 .

Below is just a sketch of the view... or a new activity altogether that can be passed to the update activity via [//workflowdata/]. All the boxes highlighted in Red do not exists _yet_.

![wal](https://github.com/microsoft/MIMWAL/assets/22840553/49212369-141c-49b4-9260-4a8a4551ccc9)

**Comment by NileshGhodekar** [comment]
Hi @ccthis, If I understood this correctly, you want to query pam requests for all users appearing in //delta/candidate/removed. I'm getting rusty on MIMWAL, but I think you should be able get all these pam requests using the [Query Collection](https://github.com/Microsoft/MIMWAL/wiki/Query-Resources) feature in conjunction with the [FormatMultivaluedList](https://github.com/Microsoft/MIMWAL/wiki/FormatMultivaluedList-Function) function...

**Reply by ccthis** [reply]
Tried with and without iteration. The second attempt is closer to _something_ resulting in a failure to enumerate. I've tried a previous activity carrying over [//Target/ObjectID], [//Delta/ObjectID], and the hardcoded ObjectID and "FormatMultivaluedList" is "failing to enumerate". Seems like a bug not being able to pass the ObjectID from parent object pamrole.

Using:
Query:
/FormatMultivaluedList("/msidmPamRequest[(Creator='{0}') and (msidmPamRequestRole='{1}') and (msidmPamRequestWasClosed = false) and (msidmPamRequestExpirationProcessedState= 'NotHandled') and (msidmPamRequestExpirationDate > fn:current-dateTime())]",[//Delta/msidmPamCandidates/Removed/ObjectID],[//Target/ObjectID])

Error Generated:
Workflow Instance '157d32ec-bb03-4eaa-a6c5-ae5ba42a180a' was aborted


Using:
Query:
/FormatMultivaluedList("/msidmPamRequest[(Creator='{0}') and (msidmPamRequestRole='{1}') and (msidmPamRequestWasClosed = false) and (msidmPamRequestExpirationProcessedState= 'NotHandled') and (msidmPamRequestExpirationDate > fn:current-dateTime())]",[//Delta/msidmPamCandidates/Removed/ObjectID],[//Delta/ObjectID])

Error Generated:
An error occurred while enumerating the filter '/msidmPamRequest[(Creator='b0959acb-1fe3-41a5-b2af-7ae33e186ee4') and (msidmPamRequestRole='') and (msidmPamRequestWasClosed = false) and (msidmPamRequestExpirationProcessedState= 'NotHandled') and (msidmPamRequestExpirationDate > fn:current-dateTime())]'.

**Comment by NileshGhodekar** [comment]
your first one looks closer to the truth. worth checking why the workflow is aborting. it may be that query is fine, but the workflow is aborting during the updates processing...

**Reply by ccthis** [reply]
Got it working, I'll share later. Need some sleep, thanks!
